src/resolve/resolved-conf.c

```diff
@@ -565,6 +662,98 @@ static void read_proc_cmdline(Manager *m) {
                 log_warning_errno(r, "Failed to read kernel command line, ignoring: %m");
 }
 
+int config_parse_dnssec_algorithms_digests(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        Manager *m = ASSERT_PTR(userdata);
+        int (*from_string)(const char*);
+        int (*init)(Manager*);
+        bool negative = false;
+        const char *object;
+        Set **set;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+
+        if (streq(lvalue, "DNSSECAlgorithms")) {
+                from_string = dnssec_algorithm_from_string;
+                init = manager_init_dnssec_algorithms;
+                set = &m->dnssec_algorithms;
+                object = "algorithm";
+        } else if (streq(lvalue, "DNSSECDigests")) {
+                from_string = dnssec_digest_from_string;
+                init = manager_init_dnssec_digests;
+                set = &m->dnssec_digests;
+                object = "digest";
+        } else
+                assert_not_reached();
+
+        r = set_ensure_allocated(set, NULL);
+        if (r < 0)
+                return r;
+
+        if (isempty(rvalue)) {
+                set_clear(*set);
+                return 0;
+        }
+
+        if (rvalue[0] == '~') {
+                negative = true;
+                rvalue++;
+        }
+
+        for (;;) {
+                _cleanup_free_ char *word = NULL;
+
+                r = extract_first_word(&rvalue, &word, NULL, EXTRACT_UNQUOTE);
+                if (r < 0) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r, "Invalid syntax, ignoring: %s", rvalue);
+                        return 0;
+                }
+                if (r == 0)
+                        break;
+
+                if (streq(word, "all")) {
+                        r = init(m);
+                        if (r < 0)
+                                return r;
+                        continue;
+                }
+
+                r = from_string(word);
+                if (r < 0) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r,
+                                   "Unknown DNSSEC %s '%s', ignoring.", object, word);
+                        continue;
+                }
+
+                if (negative) {
+                        void *e;
+                        e = set_remove(*set, INT_TO_PTR(r));
+                        if (!e)
+                                return log_syntax(unit, LOG_WARNING, filename, line, r,
+                                                  "Failed to remove DNSSEC %s '%s'.", object, word);
```

Comment:
Not `log_syntax`.

Reply:
Actually we shouldn't return on error so I will change this to normal `log_warning()`